enterprise/server/atime_updater/atime_updater.go

```diff
@@ -0,0 +1,239 @@
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+)
+
+var (
+	// The maximum gRPC message size is 4MB. Each digest proto is
+	// 64 bytes (hash) + 8 bytes (size) + 2 bytes (tags) = 74 bytes. The
+	// remaining fields in FindMissingBlobsRequest should be small (1kb?), so
+	// we can conceivably fit 4MB / 74bytes = 14.1k of these. Cap it at 10k for
+	// now and we can increase it (or increase the update frequency) if we drop
+	// a lot of updates. Note this also affects memory usage of the proxy!
+	atimeUpdaterMaxDigestsPerUpdate = flag.Int("cache_proxy.remote_atime_max_digests_per_update", 10*1000, "The maximum number of blob digests to send in a single request for updating blob access times in the remote cache.")
+	atimeUpdaterMaxUpdatesPerGroup  = flag.Int("cache_proxy.remote_atime_max_updates_per_group", 25, "The maximum number of FindMissingBlobRequests to accumulate per group for updating blob access times in the remote cache.")
+	atimeUpdaterBatchUpdateInterval = flag.Duration("cache_proxy.remote_atime_update_interval", 30*time.Second, "The time interval to wait between sending access time updates to the remote cache.")
+	atimeUpdaterRpcTimeout          = flag.Duration("cache_proxy.remote_atime_rpc_timeout", 1*time.Second, "RPC timeout to use when updating blob access times in the remote cache.")
+)
+
+// A pending batch of atime updates (basically a FindMissingBlobsRequest).
+// Stored as a struct so digest keys can be stored in a set for de-duping.
+type atimeUpdate struct {
+	instanceName   string
+	digests        map[digest.Key]struct{} // stored as a set for de-duping.
+	digestFunction repb.DigestFunction_Value
+}
+
+func (u *atimeUpdate) toProto() *repb.FindMissingBlobsRequest {
+	req := repb.FindMissingBlobsRequest{
+		InstanceName:   u.instanceName,
+		BlobDigests:    make([]*repb.Digest, len(u.digests)),
+		DigestFunction: u.digestFunction,
+	}
+	i := 0
+	for d := range u.digests {
+		req.BlobDigests[i] = d.ToDigest()
+		i++
+	}
+	return &req
+}
+
+// The set of pending access time updates to send per group. There should be at
+// most one of these per group, but each contains many atimeUpdate{}s, as
+// there is one per (groupID, instance-name, digest-function) tuple. However,
+// the update will only keep one such atimeUpdate{} and if it gets too big will
+// discard additional digests until it's flushed.
+type atimeUpdates struct {
+	mu      sync.Mutex // protects jwt, updates, and atimeUpdate.digests.
+	jwt     string
+	updates []*atimeUpdate
+}
+
+type atimeUpdater struct {
+	authenticator interfaces.Authenticator
+
+	mu      sync.Mutex               // protects updates
+	updates map[string]*atimeUpdates // pending updates keyed by groupID.
+
+	maxDigestsPerUpdate int
+	maxUpdatesPerGroup  int
+	updateInterval      time.Duration
+	rpcTimeout          time.Duration
+
+	remote repb.ContentAddressableStorageClient
+}
+
+type Opts struct {
+	TickerForTesting <-chan time.Time
```

Comment:
nit: this is also achievable using the clockwork fake clock. I would like to see us standardize on a single approach for testing behaviors that depend on time.

Reply:
Thanks for adding that! Switched all the tests here to use it and dropped the `Opts`.